File: experiments/exp1_distances/R/distances.R
# ============================================================
# File: R/distances.R
# Experiment: exp1_distances
# Description: Core statistical distance functions
# Reference: Zhang (2017) Ch.4-5; Costa et al. (2015)
# ============================================================


fisher_rao_univariate <- function(mu1, sigma1, mu2, sigma2) {
  # Fisher-Rao geodesic distance between N(mu1, sigma1^2) and N(mu2, sigma2^2).
  # Zhang (2017) eq. 4.11; Costa et al. (2015) eq. 9
  stopifnot(sigma1 > 0, sigma2 > 0)
  d_mu  <- mu1 - mu2
  F_val <- sqrt((d_mu^2 + 2 * (sigma1 - sigma2)^2) *
                (d_mu^2 + 2 * (sigma1 + sigma2)^2))
  sqrt(2) * log((F_val + d_mu^2 + 2 * (sigma1^2 + sigma2^2)) / (4 * sigma1 * sigma2))
}


fisher_rao_bivariate <- function(mu1, sigma1, mu2, sigma2) {
  # Fisher-Rao distance for bivariate Gaussians with diagonal covariance.
  # theta_k = (mu_k, c(sigma_k1, sigma_k2)); mu_k is a scalar (same for both dims).
  # Zhang (2017) eq. 4.12 / 8.2
  #
  # Numerically stable form: log((A+B)^2 / (4*s1*s2))
  # derived from A^2 - B^2 = 4*sigma1[i]*sigma2[i], avoiding cancellation when A≈B.
  stopifnot(length(sigma1) == 2, length(sigma2) == 2,
            all(sigma1 > 0), all(sigma2 > 0))
  delta_mu <- (mu1 - mu2) / sqrt(2)
  terms <- sapply(1:2, function(i) {
    A_i <- sqrt(delta_mu^2 + (sigma1[i] + sigma2[i])^2)
    B_i <- sqrt(delta_mu^2 + (sigma1[i] - sigma2[i])^2)
    log((A_i + B_i)^2 / (4 * sigma1[i] * sigma2[i]))
  })
  sqrt(2 * sum(terms^2))
}


hellinger_discrete <- function(p, q) {
  # Hellinger distance between two discrete probability distributions.
  # Zhang (2017) eq. 5.1
  stopifnot(length(p) == length(q), all(p >= 0), all(q >= 0))
  p <- p / sum(p)
  q <- q / sum(q)
  sqrt(0.5 * sum((sqrt(p) - sqrt(q))^2))
}

File: experiments/exp1_distances/python/distances.py
# ============================================================
# File: python/distances.py
# Experiment: exp1_distances
# Description: Core statistical distance functions
# Reference: Zhang (2017) Ch.4-5; Costa et al. (2015)
# ============================================================

import numpy as np


def fisher_rao_univariate(mu1: float, sigma1: float, mu2: float, sigma2: float) -> float:
    """Fisher-Rao geodesic distance between N(mu1, sigma1^2) and N(mu2, sigma2^2).

    Zhang (2017) eq. 4.11; Costa et al. (2015) eq. 9
    """
    assert sigma1 > 0 and sigma2 > 0, "sigma must be positive"
    d_mu = mu1 - mu2
    F = np.sqrt(
        (d_mu**2 + 2 * (sigma1 - sigma2)**2) *
        (d_mu**2 + 2 * (sigma1 + sigma2)**2)
    )
    return float(
        np.sqrt(2) * np.log((F + d_mu**2 + 2 * (sigma1**2 + sigma2**2)) / (4 * sigma1 * sigma2))
    )


def fisher_rao_bivariate(
    mu1: float, sigma1: np.ndarray,
    mu2: float, sigma2: np.ndarray,
) -> float:
    """Fisher-Rao distance for bivariate Gaussians with diagonal covariance.

    theta_k = (mu_k, [sigma_k1, sigma_k2]); mu_k is a scalar (same for both dims).
    Numerically stable form: log((A+B)^2 / (4*s1*s2)),
    derived from A^2 - B^2 = 4*sigma1[i]*sigma2[i], avoiding cancellation when A≈B.
    Zhang (2017) eq. 4.12 / 8.2
    """
    sigma1 = np.asarray(sigma1, dtype=float)
    sigma2 = np.asarray(sigma2, dtype=float)
    assert sigma1.shape == (2,) and sigma2.shape == (2,), "sigma must be length-2 arrays"
    assert np.all(sigma1 > 0) and np.all(sigma2 > 0), "sigma must be positive"

    delta_mu = (mu1 - mu2) / np.sqrt(2)
    terms = np.zeros(2)
    for i in range(2):
        A_i = np.sqrt(delta_mu**2 + (sigma1[i] + sigma2[i])**2)
        B_i = np.sqrt(delta_mu**2 + (sigma1[i] - sigma2[i])**2)
        terms[i] = np.log((A_i + B_i)**2 / (4 * sigma1[i] * sigma2[i]))
    return float(np.sqrt(2 * np.sum(terms**2)))


def hellinger_discrete(p: np.ndarray, q: np.ndarray) -> float:
    """Hellinger distance between two discrete probability distributions.

    Zhang (2017) eq. 5.1
    """
    p = np.asarray(p, dtype=float)
    q = np.asarray(q, dtype=float)
    assert len(p) == len(q), "p and q must have the same length"
    assert np.all(p >= 0) and np.all(q >= 0), "probabilities must be non-negative"
    p = p / p.sum()
    q = q / q.sum()
    return float(np.sqrt(0.5 * np.sum((np.sqrt(p) - np.sqrt(q))**2)))
